crates/rethnet_rpc_server/src/lib.rs

```diff
@@ -1 +1,200 @@
+                            }
+                            _ => { // TODO: after adding all the methods here, eliminate this
+                                   // catch-all match arm.
+                                Json(serde_json::json!(jsonrpc::Response {
+                                    jsonrpc: jsonrpc::Version::V2_0,
+                                    id,
+                                    data: jsonrpc::ResponseData::<serde_json::Value>::Error { error: jsonrpc::Error {
+                                        code: 0,
+                                        message: String::from("unsupported JSON-RPC method"),
+                                        data: Some(serde_json::to_value(method).unwrap()),
+                                    }}
+                                }))
+                            }
+                        }
+                    }
+                }
+            }
+        )
+    )
+    .with_state(state)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use axum::{body::Body, http::Request};
+    use rethnet_eth::{
+        remote::{jsonrpc, BlockSpec, MethodInvocation},
+        Address,
+    };
+    use rethnet_evm::KECCAK_EMPTY;
+    use revm::primitives::state::AccountInfo;
+    use std::net::{SocketAddr, TcpListener};
+
+    fn start_server() -> SocketAddr {
+        let listener = TcpListener::bind("0.0.0.0:0".parse::<SocketAddr>().unwrap()).unwrap();
+        let address = listener.local_addr().unwrap();
+
+        let mut accounts: hashbrown::HashMap<Address, AccountInfo> = Default::default();
+        accounts.insert(
+            Address::from_low_u64_ne(1),
+            AccountInfo {
+                code: None,
+                code_hash: KECCAK_EMPTY,
+                balance: U256::ZERO,
+                nonce: 0,
+            },
+        );
+
+        let rethnet_state = Arc::new(Mutex::new(HybridState::with_accounts(accounts)));
+
+        tokio::spawn(async move {
+            axum::Server::from_tcp(listener)
+                .unwrap()
+                .serve(router(rethnet_state).await.into_make_service())
+                .await
+                .unwrap();
+        });
+
+        address
+    }
+
+    async fn submit_request(address: &SocketAddr, request: &RpcRequest) -> String {
+        String::from_utf8(
+            hyper::body::to_bytes(
+                hyper::Client::new()
+                    .request(
+                        Request::post(format!("http://{address}/"))
+                            .header("Content-Type", "application/json")
+                            .body(Body::from(serde_json::to_string(&request).unwrap()))
+                            .unwrap(),
+                    )
+                    .await
+                    .unwrap()
+                    .into_body(),
+            )
+            .await
+            .unwrap()
+            .to_vec(),
+        )
+        .expect("should decode response from UTF-8")
+    }
+
+    #[tokio::test]
```

Comment:
Given that these are integration tests - using the public API of rethnet_rpc_server - you can use the integration test folder: `crates/rethnet_rpc_server/tests`

See https://doc.rust-lang.org/cargo/guide/project-layout.html

Reply:
but, _are_ these integration tests? i'm happy to defer to your judgement, but these actually felt like unit tests to me, and that's why i kept them here. i know that sometimes people say things like, "if it's doing network interaction then it's an integration test." but to me, an integration test is a test that exercises the integration of multiple different components. and to me that seems to be the reason for having the integration test folder in the rust project hierarchy: it provides a place to put tests that combine multiple different components, because such tests would be out of place within the `mod tests` sub-module of a single component, because they're not specific to any one component. here, however, all of the tests are specific to the contents of this `lib` module, so i felt they belonged within that module. again, i'm happy to defer to your preference, but that was my rationale.

Reply:
I generally agree with what you said, but based on your description I'd recognise four components at work here:

1. Serialization
2. Client (reqwest)
3. Server (listener + handler)
4. Deserialization

As such, I'd say that it qualifies as an integration test under your definition too.

The Rust doc says something along similar lines: https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html

Reply:
addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/37b288e819d30d5913a8cb7425bc47657ec59912